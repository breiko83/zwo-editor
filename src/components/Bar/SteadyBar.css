.segment {
  position: relative;
}
.bar {    
  border-radius: 10px;
  border: 1px solid white;
}

.comment {
  position: absolute;
  top: -60px;
  left: 105px;
  color: gray;
  padding: 5px;  
  font-size: 20px;
}