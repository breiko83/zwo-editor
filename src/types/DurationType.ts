export type DurationType = "time" | "distance";