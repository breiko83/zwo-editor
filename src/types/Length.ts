export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;
