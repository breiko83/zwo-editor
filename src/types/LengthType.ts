export type LengthType = "time" | "distance";