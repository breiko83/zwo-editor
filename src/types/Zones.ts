export enum ZoneColor {
  GRAY = '#807F80',
  BLUE = '#0E90D4',
  GREEN = '#00C46A',
  YELLOW = '#FFCB00',
  ORANGE = '#FF6430',
  RED = '#E90000',
  WHITE = '#FFFFFF',
}

export const Zones = {
  Z1: {
    min: 0.1,
    max: 0.605
  },
  Z2: {
    min: 0.605,
    max: 0.763
  },
  Z3: {
    min: 0.763,
    max: 0.901
  },
  Z4: {
    min: 0.901,
    max: 1.053
  },
  Z5: {
    min: 1.053,
    max: 1.194
  },
  Z6: {
    min: 1.194,
    max: 2.0
  }
}

export const ZonesArray = [
  [Zones.Z1.min,Zones.Z1.max],
  [Zones.Z2.min,Zones.Z2.max],
  [Zones.Z3.min,Zones.Z3.max],
  [Zones.Z4.min,Zones.Z4.max],
  [Zones.Z5.min,Zones.Z5.max],
  [Zones.Z6.min,Zones.Z6.max],
]

export const zoneColor = (intensity: number): ZoneColor => {
  if (intensity < Zones.Z1.max) {
    return ZoneColor.GRAY;
  } else if (intensity < Zones.Z2.max) {
    return ZoneColor.BLUE;
  } else if (intensity < Zones.Z3.max) {
    return ZoneColor.GREEN;
  } else if (intensity < Zones.Z4.max) {
    return ZoneColor.YELLOW;
  } else if (intensity < Zones.Z5.max) {
    return ZoneColor.ORANGE;
  } else {
    return ZoneColor.RED;
  }
};
