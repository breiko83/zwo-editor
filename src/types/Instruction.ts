import { v4 as uuidv4 } from "uuid";
import { WorkoutMode } from "../modes/WorkoutMode";
import { Distance, Duration, Length } from "./Length";

export interface Instruction {
  id: string,
  text: string,
  offset: Length,
}

export function createInstruction(fields: Partial<Instruction>, mode: WorkoutMode): Instruction {
  return {
    id: uuidv4(),
    text: "",
    offset: mode.lengthType === "time" ? new Duration(0) : new Distance(0),
    ...fields,
  };
}
