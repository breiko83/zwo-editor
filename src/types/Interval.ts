import { Length } from "./Length";
import { PaceType } from "./PaceType";

export interface SteadyInterval {
  id: string,
  length: Length,
  type: 'steady',
  intensity: number,
  cadence: number,
  pace: PaceType,
}

export interface RampInterval {
  id: string,
  length: Length,
  type: 'ramp',
  startIntensity: number,
  endIntensity: number,
  cadence: number,
  pace: PaceType,
}

export interface FreeInterval {
  id: string,
  length: Length,
  type: 'free',
  cadence: number,
}

export interface RepetitionInterval {
  id: string,
  type: 'repetition',
  onCadence: number,
  offCadence: number,
  onIntensity: number,
  offIntensity: number,
  onLength: Length,
  offLength: Length,
  repeat: number,
  pace: PaceType,
}

export type Interval = SteadyInterval | RampInterval | FreeInterval | RepetitionInterval;
