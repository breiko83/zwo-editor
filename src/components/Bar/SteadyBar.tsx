import React, { useState, useEffect } from 'react'
import './SteadyBar.css'
import { zoneColor, Zones } from '../../types/Zones'
import { Resizable } from 're-resizable'
import Label from '../Label/Label'
import { SteadyInterval } from '../../types/Interval'
import { intensityMultiplier } from './multipliers'
import { WorkoutMode } from '../../modes/WorkoutMode'

interface SteadyBarProps {
  interval: SteadyInterval;
  mode: WorkoutMode;
  onChange: (interval: SteadyInterval) => void;
  onClick: (id: string) => void;
  selected: boolean;
  showLabel: boolean;
}

const SteadyBar = ({interval, mode, ...props}: SteadyBarProps) => {
  const style = { backgroundColor: zoneColor(interval.intensity) }

  const [width, setWidth] = useState(mode.lengthToWidth(interval.length))

  const [height, setHeight] = useState(mode.intensityToHeight(interval.intensity))

  const [showLabel, setShowLabel] = useState(false)

  const [selected, setSelected] = useState(props.selected)

  useEffect(()=>{
    setSelected(props.selected) 
  },[props.selected])  

  const handleCadenceChange = (cadence: number) => {
    props.onChange({ ...interval, cadence })
  }

  const handleResizeStop = (dWidth: number, dHeight: number) => {
    setWidth(width + dWidth)
    setHeight(height + dHeight)

    notifyChange(dWidth, dHeight)
  }

  const notifyChange = (dWidth: number, dHeight: number) => {
    props.onChange({
      ...interval,
      length: mode.widthToLength(width + dWidth),
      intensity: mode.heightToIntensity(height + dHeight),
    })
  }

  return (
    <div className='segment'
      onMouseEnter={() => setShowLabel(true)}
      onMouseLeave={() => setShowLabel(false)}
      onClick={() => props.onClick(interval.id)}
      style={props.selected ? { zIndex: 10 } : {}}
    >
      {((selected || showLabel) && (props.showLabel)) &&
        <Label
          interval={interval}
          mode={mode}
          onCadenceChange={(cadence: number)=> handleCadenceChange(cadence)}
        />
      }
      <Resizable
        className='bar'
        size={{
          width: mode.lengthToWidth(interval.length),
          height: mode.intensityToHeight(interval.intensity),
        }}
        minWidth={3}
        minHeight={intensityMultiplier * Zones.Z1.min}
        maxHeight={intensityMultiplier * Zones.Z6.max}
        enable={{ top: true, right: true }}
        grid={[1, 1]}
        onResizeStop={(e, direction, ref, d) => handleResizeStop(d.width, d.height)}
        onResize={(e, direction, ref, d) => notifyChange(d.width, d.height)}
        style={style}
      >
      </Resizable>
    </div>

  );
}

export default SteadyBar