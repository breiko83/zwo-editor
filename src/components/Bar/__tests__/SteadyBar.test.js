import React from 'react';
import SteadyBar from '../SteadyBar';
import { Zones } from '../../../types/Zones'
import renderer from 'react-test-renderer';
import { v4 as uuidv4 } from 'uuid'
import '@testing-library/jest-dom/extend-expect'

test('SteadyBar renders correctly', () => {
  const interval = {
    duration: 50,
    power: Zones.Z3.min,
    type: 'steady',
    id: uuidv4()
  }

  const ftp = 250

  const component = renderer.create(
    <SteadyBar
      interval={interval}
      ftp={ftp}
      onChange={() => { }}
      onClick={() => { }}
    />
  )

  expect(component).toMatchSnapshot();
})