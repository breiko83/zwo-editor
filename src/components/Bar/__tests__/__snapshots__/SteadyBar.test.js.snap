// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SteadyBar renders correctly 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="bar"
    style={
      Object {
        "backgroundColor": "#00C46A",
        "boxSizing": "border-box",
        "flexShrink": 0,
        "height": "190.75px",
        "maxHeight": 500,
        "maxWidth": undefined,
        "minHeight": 25,
        "minWidth": 3,
        "position": "relative",
        "userSelect": "auto",
        "width": "0px",
      }
    }
  >
    <span>
      <div
        className=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "cursor": "row-resize",
            "height": "10px",
            "left": "0px",
            "position": "absolute",
            "top": "-5px",
            "userSelect": "none",
            "width": "100%",
          }
        }
      />
      <div
        className=""
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "cursor": "col-resize",
            "height": "100%",
            "position": "absolute",
            "right": "-5px",
            "top": "0px",
            "userSelect": "none",
            "width": "10px",
          }
        }
      />
    </span>
  </div>
</div>
`;
